File: api/schedule.go
package handler

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/SumitNalavade/FRISCOISDHACAPIV2/utils"
)

func ScheduleHandler(w http.ResponseWriter, r *http.Request) {
	var courses []utils.StudentScheduleType

	queryParams := r.URL.Query()

	username := queryParams.Get("username")
	password := queryParams.Get("password")

	if username == "john" && password == "doe" {
		response, _ := json.Marshal(utils.FakeStudentSchedule)
		w.Header().Add("Content-Type", "application/json")
		fmt.Fprint(w, string(response))
		return
	}

	pageContent := utils.GetPageContent(username, password, utils.SCHEDULEURL)
	doc, _ := goquery.NewDocumentFromReader(strings.NewReader(pageContent))

	doc.Find("tr.sg-asp-table-data-row").Each(func(i int, s *goquery.Selection) {
		var newCourse utils.StudentScheduleType

		s.Find("td").Each(func(i int, s *goquery.Selection) {
			text := strings.TrimSpace(s.Text())

			switch i {
			case 0:
				newCourse.Code = text
			case 1:
				newCourse.Name = text
			case 2:
				newCourse.Periods = text
			case 3:
				newCourse.Teacher = text
			case 4:
				newCourse.Room = text
			case 5:
				newCourse.Days = text
			case 6:
				newCourse.MarkingPeriods = text
			case 7:
				newCourse.Building = text
			case 8:
				newCourse.Status = text
			}
		})

		courses = append(courses, newCourse)
	})

	response, _ := json.Marshal(courses)
	
	if(string(response) == "null") {
		w.WriteHeader(http.StatusUnauthorized)
		fmt.Fprint(w, "Something went wrong!")
		return
	}

	w.Header().Add("Content-Type", "application/json")
	fmt.Fprint(w, string(response))
}

File: api/schedule.py
from http.server import BaseHTTPRequestHandler
from urllib import parse
import json

from bs4 import BeautifulSoup
from api._lib.getRequestSession import getRequestSession


SCHEDULE_URL = "https://hac.friscoisd.org/HomeAccess/Content/Student/Classes.aspx"

def get_or_none(tds, i):
    return tds[i] if i < len(tds) else None

class handler(BaseHTTPRequestHandler):

    def do_GET(self):
        query = parse.urlsplit(self.path).query
        params = dict(parse.parse_qsl(query))

        username = params.get("username")
        password = params.get("password")

        if not username or not password:
            return self._send_json(
                {"error": "Missing username or password in query parameters."},
                status=400,
            )

        try:
            session = getRequestSession(username, password)

            resp = session.get(SCHEDULE_URL)
            resp.raise_for_status()
            html = resp.text

            soup = BeautifulSoup(html, "lxml")

            schedule = []
            # Find all schedule rows in the main schedule table
            for row in soup.select("#plnMain_dgSchedule > tr.sg-asp-table-data-row"):
                tds = [td.get_text(strip=True) for td in row.find_all("td")]

                schedule.append({
                    "building":       get_or_none(tds, 7),
                    "courseCode":     get_or_none(tds, 0),
                    "courseName":     get_or_none(tds, 1),
                    "days":           get_or_none(tds, 5),
                    "markingPeriods": get_or_none(tds, 6),
                    "periods":        get_or_none(tds, 2),
                    "room":           get_or_none(tds, 4),
                    "status":         get_or_none(tds, 8),
                    "teacher":        get_or_none(tds, 3),
                })

            return self._send_json({"studentSchedule": schedule}, status=200)

        except Exception:
            return self._send_json(
                {"studentSchedule": [], "error": "Failed to fetch schedule."},
                status=500,
            )

    def _send_json(self, payload, status=200):
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)
